src/sdam/state/mod.rs

```diff
@@ -2,161 +2,302 @@ pub(super) mod server;
+    /// Creates a new server selection timeout error message given the `criteria`.
+    pub(crate) async fn server_selection_timeout_error_message(
+        &self,
+        criteria: &SelectionCriteria,
+    ) -> String {
+        self.state
+            .read()
+            .await
+            .description
+            .server_selection_timeout_error_message(criteria)
     }
 
+    /// Signals the SDAM background threads that they should wake up and check the topology.
     pub(crate) fn request_topology_check(&self) {
-        for server in self.servers.values() {
-            server.request_topology_check();
+        self.common.message_manager.request_topology_check();
+    }
+
+    /// Waits until either the `timeout` has elapsed or a topology check has been requested.
+    pub(crate) async fn wait_for_topology_check_request(&self, timeout: Duration) -> bool {
+        self.common
+            .message_manager
+            .wait_for_topology_check_request(timeout)
+            .await
+    }
+
+    /// Wakes all tasks waiting for a topology change.
+    pub(crate) fn notify_topology_changed(&self) {
+        self.common.message_manager.notify_topology_changed();
+    }
+
+    /// Waits until either the `timeout` has elapsed or the topology has been updated.
+    pub(crate) async fn wait_for_topology_change(&self, timeout: Duration) -> bool {
+        self.common
+            .message_manager
+            .wait_for_topology_change(timeout)
+            .await
+    }
+
+    /// Handles an error that occurs before the handshake has completed during an operation.
+    pub(crate) async fn handle_pre_handshake_error(&self, error: Error, address: StreamAddress) {
+        if error.is_network_error() {
+            self.mark_server_as_unknown(error, address).await;
         }
     }
 
-    fn add_new_server(
-        &mut self,
-        address: StreamAddress,
-        wrapped_topology: &Arc<RwLock<Topology>>,
-    ) -> Result<()> {
+    /// Handles an error that occurs after the handshake has completed during an operation.
+    pub(crate) async fn handle_post_handshake_error(
+        &self,
+        error: Error,
+        conn: Connection,
+        server: Arc<Server>,
+    ) {
+        // If we encounter certain errors, we must update the topology as per the
+        // SDAM spec.
+        if error.is_non_timeout_network_error() {
+            self.mark_server_as_unknown(error, server.address.clone())
+                .await;
+            server.clear_connection_pool();
+        } else if error.is_recovering() || error.is_not_master() {
+            self.mark_server_as_unknown(error.clone(), server.address.clone())
+                .await;
+
+            self.common.message_manager.request_topology_check();
+
+            let wire_version = conn
+                .stream_description()
+                .map(|sd| sd.max_wire_version)
+                .ok()
+                .and_then(std::convert::identity)
```

Comment:
I think we can use `flatten` for this now.

Reply:
Sounds good; in that case, we'll also need to update the minimum version to 1.40 before we release, so we could also add the `nonexhaustive` attribute to our enums.